src/hotspot/share/gc/g1/g1ParScanThreadState.cpp

```diff
@@ -207,7 +207,7 @@ void G1ParScanThreadState::do_oop_evac(T* p) {
 
   markWord m = obj->mark();
   if (m.is_marked()) {
-    obj = cast_to_oop(m.decode_pointer());
+    obj = obj->forwardee(m);
```

Comment:
Shall we have a method "oop::forwardee_not_self" which guarantee to be not self fowarded? So we can remove the self-forward if-check in GC critical path.

Reply:
Are there any paths that don't need to handle self-forwarded state?

Also, it seems to me that the path would be dominated by the load of the mark-word. Testing-and-branching for the self-forwarded bit seems like the minor problem there. It would be nice if we could tell the C++ compiler that the branch is expected to be uncommon, so it could shape the emitted code accordingly, but, afaik, we can't.

If we were to micro-optimize the forwarding-decoding, then it would be more useful to optimize the common pattern:

```
if (o->is_forwarded()) { // Loads and tests the mark-word
  oop fwd = o->forwardee(); // Loads mark-word again, and decode forwardee.
 ...
}
```

To something like:

```
oop fwd = o->forwardee(); // Return nullptr when not forwarded
if (fwd != nullptr) {
  ...
}
```

There's a way to improve further on this, as I proposed back in the early versions of #5955, which also avoids the decoding altogether if not forwarded, but it's a little more clunky:

```
OopForwarding fwd(obj);
if (fwd.is_forwarded()) {
  oop forwardee = fwd.forwardee();
  ...
}
```

where the scoped OopForwarding object would encapsulate the markWord and the testing and decoding of the fwd-ptr, but it has been rejected back then. But it would certainly help more than an oopDesc::forwardee_not_self() approach (if that were even possible, which I think it is not).

Reply:
Sorry. It's my misunderstanding. G1ParScanThreadState::do_oop_evac still needs to handle self forward.